linkerd/app/outbound/src/tcp/logical.rs

```diff
@@ -127,3 +110,223 @@ where
+        let ep1_addr = SocketAddr::new([192, 0, 2, 31].into(), 3333);
+        let resolve = support::resolver();
+        let resolved = resolve.handle();
+        let mut resolve_tx = resolve.endpoint_tx(logical_addr);
+
+        // Build the TCP logical stack with a mocked endpoint stack that alters its response stream
+        // based on the address.
+        let (rt, _shutdown) = runtime();
+        let svc = Outbound::new(default_config(), rt)
+            .with_stack(svc::mk(move |ep: Endpoint| match ep.addr {
+                Remote(ServerAddr(addr)) if addr == ep0_addr => {
+                    tracing::debug!(%addr, "writing ep0");
+                    let mut io = support::io();
+                    io.write(b"who r u?").read(b"ep0");
+                    future::ok::<_, support::io::Error>(io.build())
+                }
+                Remote(ServerAddr(addr)) if addr == ep1_addr => {
+                    tracing::debug!(%addr, "writing ep1");
+                    let mut io = support::io();
+                    io.write(b"who r u?").read(b"ep1");
+                    future::ok::<_, support::io::Error>(io.build())
+                }
+                addr => unreachable!("unexpected endpoint: {}", addr),
+            }))
+            .push_tcp_logical(resolve)
+            .into_inner()
+            .new_service(logical);
+
+        // We add a single endpoint to the balancer and it is used:
+
+        resolve_tx
+            .add(Some((ep0_addr, Default::default())))
+            .unwrap();
+        time::sleep(time::Duration::from_secs(1)).await; // Let the balancer observe the update.
+        let (io, task) = spawn_io();
+        svc.clone().oneshot(io).await.unwrap();
+        let msg = task.await.unwrap().unwrap();
+        assert_eq!(msg, "ep0");
+
+        // When we add a second endpoint, traffic is sent to both endpoints:
+
+        resolve_tx
+            .add(Some((ep1_addr, Default::default())))
+            .unwrap();
+        time::sleep(time::Duration::from_secs(1)).await; // Let the balancer observe the update.
+        let mut seen0 = false;
+        let mut seen1 = false;
+        for i in 1..=100 {
+            let (io, task) = spawn_io();
+            svc.clone().oneshot(io).await.unwrap();
+            let msg = task.await.unwrap().unwrap();
+            match msg.as_str() {
+                "ep0" => {
+                    seen0 = true;
+                }
+                "ep1" => {
+                    seen1 = true;
+                }
+                msg => unreachable!("unexpected read: {}", msg),
+            }
+            assert!(resolved.only_configured(), "Resolution must be reused");
+            if seen0 && seen1 {
+                tracing::info!("Both endpoints observed after {} iters", i);
+                break;
+            }
+            if i % 10 == 0 {
+                tracing::debug!("iters={} ep0={} ep1={}", i, seen0, seen1);
```

Comment:
nit, doesn't matter:
```suggestion
                tracing::debug!(iters = i, ep0 = seen0, ep1 = seen1);
```